src/3d/qgsshadowrenderingframegraph.cpp

```diff
@@ -0,0 +1,342 @@
+
+  mShadowClearBuffers = new Qt3DRender::QClearBuffers( mShadowRenderTargetSelector );
+  mShadowClearBuffers ->setBuffers( Qt3DRender::QClearBuffers::BufferType::ColorDepthBuffer );
+
+  mShadowRenderStateSet = new Qt3DRender::QRenderStateSet( mShadowClearBuffers );
+  mShadowDepthTest = new Qt3DRender::QDepthTest;
+  mShadowDepthTest->setDepthFunction( Qt3DRender::QDepthTest::Less );
+  mShadowRenderStateSet->addRenderState( mShadowDepthTest );
+  mShadowCullFace = new Qt3DRender::QCullFace;
+  mShadowCullFace->setMode( Qt3DRender::QCullFace::NoCulling );
+  mShadowRenderStateSet->addRenderState( mShadowCullFace );
+
+  return mShadowSceneEntitiesFilter;
+}
+
+Qt3DRender::QFrameGraphNode *QgsShadowRenderingFrameGraph::constructPostprocessingPass()
+{
+  mPostprocessPassLayerFilter = new Qt3DRender::QLayerFilter;
+  mPostprocessPassLayerFilter->addLayer( mPostprocessPassLayer );
+
+  mPostprocessClearBuffers = new Qt3DRender::QClearBuffers( mPostprocessPassLayerFilter );
+  mPostprocessClearBuffers->setClearColor( QColor::fromRgbF( 1.0f, 0.0f, 0.0f ) );
+
+  return mPostprocessPassLayerFilter;
+}
+
+QgsShadowRenderingFrameGraph::QgsShadowRenderingFrameGraph( QWindow *window, Qt3DRender::QCamera *mainCamera, Qt3DCore::QEntity *root )
+{
+  mRootEntity = root;
+  mMainCamera = mainCamera;
+  mLightCamera = new Qt3DRender::QCamera;
+
+  mPostprocessPassLayer = new Qt3DRender::QLayer;
+  mPreviewLayer = new Qt3DRender::QLayer;
+  mCastShadowsLayer = new Qt3DRender::QLayer;
+  mForwardRenderLayer = new Qt3DRender::QLayer;
+
+#if QT_VERSION >= QT_VERSION_CHECK(5, 10, 0)
+  mPostprocessPassLayer->setRecursive( true );
+  mPreviewLayer->setRecursive( true );
+  mCastShadowsLayer->setRecursive( true );
+  mForwardRenderLayer->setRecursive( true );
+#endif
+
+  mRenderSurfaceSelector = new Qt3DRender::QRenderSurfaceSelector;
+  mRenderSurfaceSelector->setSurface( window );
+
+  mMainViewPort = new Qt3DRender::QViewport( mRenderSurfaceSelector );
+  mMainViewPort->setNormalizedRect( QRectF( 0.0f, 0.0f, 1.0f, 1.0f ) );
+
+  mMainCameraSelector = new Qt3DRender::QCameraSelector( mMainViewPort );
+  mMainCameraSelector->setCamera( mMainCamera );
+
+  // Forward render
+  Qt3DRender::QFrameGraphNode *forwardRenderPass = constructForwardRenderPass();
+  forwardRenderPass->setParent( mMainCameraSelector );
+
+  // shadow rendering pass
+  mLightCameraSelector = new Qt3DRender::QCameraSelector( mMainViewPort );
+  mLightCameraSelector->setCamera( mLightCamera );
+
+  Qt3DRender::QFrameGraphNode *shadowRenderPass = constructShadowRenderPass();
+  shadowRenderPass->setParent( mLightCameraSelector );
+
+  // post process
+  Qt3DRender::QFrameGraphNode *postprocessingPass = constructPostprocessingPass();
+  postprocessingPass->setParent( mLightCameraSelector );
+
+  QString vertexShaderPath = QStringLiteral( "qrc:/shaders/postprocess.vert" );
+  QString fragmentShaderPath = QStringLiteral( "qrc:/shaders/postprocess.frag" );
```

Comment:
I would suggest that QgsPostprocessingEntity should not take paths to shaders in constructor - that should be an implementation detail within that entity and its material